tscat_gui/utils/keyword_list.py

```diff
@@ -0,0 +1,172 @@
+
+
+class _Keyword(QtWidgets.QFrame):
+    updated = QtCore.Signal(QtWidgets.QWidget)
+    closed = QtCore.Signal(QtWidgets.QWidget)
+
+    class Label(QtWidgets.QLabel):
+        clicked = QtCore.Signal()
+
+        def __init__(self, text: str, parent=None):
+            super().__init__(text, parent)
+
+        def mousePressEvent(self, event: QtGui.QMouseEvent) -> None:
+            if event.button() == QtCore.Qt.MouseButton.LeftButton:
+                self.clicked.emit()
+            super().mousePressEvent(event)
+
+    class DeleteLabel(QtWidgets.QLabel):
+        clicked = QtCore.Signal()
+
+        def __init__(self, parent=None):
+            super().__init__('✕', parent)
+            self.setMargin(1)
+            self.setObjectName('DeleteLabel')
+            self.setStyleSheet(
+                "#DeleteLabel { border-radius: 5px; background: #444; min-width: 10px; min-height: 10px; }")
+
+        def enterEvent(self, event: QtCore.QEvent) -> None:
+            self.setText('✖')
+            super().enterEvent(event)
+
+        def leaveEvent(self, event: QtCore.QEvent) -> None:
+            self.setText('✕')
+            super().leaveEvent(event)
+
+        def mousePressEvent(self, event: QtGui.QMouseEvent) -> None:
+            if event.button() == QtCore.Qt.MouseButton.LeftButton:
+                self.clicked.emit()
+
+            super().mousePressEvent(event)
+
+    def __init__(self, text: str, tag_list: QtCore.QObject, parent=None):
+        super().__init__(parent)
+
+        self.text = text
+        self.tag_list = tag_list
+
+        self.setObjectName('OneKeyword')
+        self.setStyleSheet("#OneKeyword { border-radius: 15px; background: #999; }")
+
+        self.delete_label = _Keyword.DeleteLabel()
+        self.delete_label.clicked.connect(lambda: self.closed.emit(self))
+
+        self.layout = QtWidgets.QHBoxLayout()
+        self.layout.addWidget(self.delete_label)
+        self.layout.setMargin(5)
+        self.setLayout(self.layout)
+
+        self.lineedit = None
+        self.tag_label = None
+        self.display()
+
+    def display(self):
+        if self.tag_label:
+            return
+
+        if self.lineedit:
+            self.layout.removeWidget(self.lineedit)
+            self.lineedit.deleteLater()
+            self.lineedit = None
+
+        self.tag_label = _Keyword.Label(self.text)
+        self.tag_label.clicked.connect(lambda: self.tag_list._request_edit(self))
+        self.layout.insertWidget(0, self.tag_label)
+
+    def edit(self):
+        if self.lineedit:
+            return
+
+        if self.tag_label:
+            self.layout.removeWidget(self.tag_label)
+            self.tag_label.deleteLater()
+            self.tag_label = None
+
+        completer = QtWidgets.QCompleter(self.tag_list.completion_strings)
+        completer.setCaseSensitivity(QtCore.Qt.CaseInsensitive)
+
+        self.lineedit = QtWidgets.QLineEdit(self.text)
```

Comment:
Did you check that we have no leaks here?

Reply:
Other by reading the code, no. What else do you have in mind?